infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/CorfuReplicationManager.java

```diff
@@ -40,83 +36,100 @@ public class CorfuReplicationManager {
-        for (ReplicationSubscriber subscriber : context.getConfig().getReplicationSubscriberToStreamsMap().keySet()) {
-            try {
-                startLogReplicationRuntime(remoteCluster, new ReplicationSession(remoteCluster.getClusterId(),
-                    subscriber));
-            } catch (Exception e) {
-                log.error("Failed to start log replication runtime for remote session {}, replication model {}, " +
-                    "client {}", remoteCluster.getClusterId(), subscriber.getReplicationModel(),
-                    subscriber.getClient());
-            }
+    public void start(ClusterDescriptor remoteCluster, LogReplicationSession session, LogReplicationContext context) {
+        try {
+            // TODO (V2): we might think of unifying the info in ClusterDescriptor into session (all nodes host+port)
+            startLogReplicationRuntime(remoteCluster, session, context);
+        } catch (Exception e) {
+            log.error("Failed to start log replication runtime for session={}", session);
         }
     }
 
     /**
-     * Stop log replication for all the sink sites
+     * Stop log replication for all sessions
      */
     public void stop() {
-        runtimeToRemoteSession.values().forEach(runtime -> {
-            try {
-                log.info("Stop log replication runtime to remote cluster id={}", runtime.getRemoteClusterId());
-                runtime.stop();
-            } catch (Exception e) {
-                log.warn("Failed to stop log replication runtime to remote cluster id={}", runtime.getRemoteClusterId());
-            }
+        sessionRuntimeMap.keySet().forEach(session -> {
+            log.info("Stop log replication runtime to remote cluster id={}", session.getSinkClusterId());
+            stopLogReplicationRuntime(session);
         });
-        runtimeToRemoteSession.clear();
+        sessionRuntimeMap.clear();
     }
 
     /**
-     * Start Log Replication Runtime to a specific Sink Session
+     * Stop log replication for specific sessions
+     *
+     * @param sessions
      */
+    public void stop(Set<LogReplicationSession> sessions) {
+        sessions.forEach(session -> {
+            stopLogReplicationRuntime(session);
+            sessionRuntimeMap.remove(session);
+        });
+    }
+
     private void startLogReplicationRuntime(ClusterDescriptor remoteClusterDescriptor,
-                                            ReplicationSession replicationSession) {
+                                            LogReplicationSession session, LogReplicationContext context) {
         try {
-            if (!runtimeToRemoteSession.containsKey(replicationSession)) {
-                log.info("Starting Log Replication Runtime for session {}", replicationSession);
-                connect(remoteClusterDescriptor, replicationSession);
+            if (!sessionRuntimeMap.containsKey(session)) {
```

Comment:
sessionRuntimeMap should contain the session, right? I see a negation sign in the check. If this is expected, can have a comment explaining this?

Reply:
Please ignore this.